lcls_tools/common/devices/screen/screen.py

```diff
@@ -0,0 +1,209 @@
+import datetime
+import os
+from typing import (
+    Any,
+    Dict,
+    Optional,
+)
+from threading import Thread
+
+from lcls_tools.common.devices.device import (
+    Device,
+    ControlInformation,
+    Metadata,
+    PVSet,
+)
+
+from epics import PV
+import h5py
+from pydantic import (
+    BaseModel,
+    SerializeAsAny,
+    field_validator,
+)
+import numpy as np
+
+
+class ScreenPVSet(PVSet):
+    """
+    The PV interface for screens is not uniform.
+    We list the potential PVs below and only
+    use the ones that are set to be PV-typed after
+    initialisation.
+    """
+
+    image: PV
+    n_col: PV
+    n_row: PV
+    n_bits: PV
+    resolution: PV
+
+    def __init__(self, **kwargs):
+        super().__init__(**kwargs)
+
+    @field_validator("*", mode="before")
+    def validate_pv_fields(cls, v: str):
+        if v:
+            return PV(v)
+
+
+class ScreenControlInformation(ControlInformation):
+    PVs: SerializeAsAny[ScreenPVSet]
+
+    def __init__(self, **kwargs):
+        super().__init__(**kwargs)
+
+
+class Screen(Device):
+    controls_information: SerializeAsAny[ScreenControlInformation]
+    metadata: SerializeAsAny[Metadata]
+    saving_images: Optional[bool] = False
+
+    def __init__(self, **kwargs):
+        super().__init__(**kwargs)
+        self._root_hdf5_location: Optional[str] = "."
+        self._last_save_filepath: Optional[str] = ""
+        self.saving_images = False
+
+    @property
+    def image(self):
+        return self.controls_information.PVs.image.get(as_numpy=True).reshape(
+            self.n_rows, self.n_columns
+        )
+
+    @property
+    def image_timestamp(self):
+        """get last timestamp for last PV activity"""
+        return self.controls_information.PVs.image.timestamp
+
+    @property
+    def hdf_save_location(self):
+        return self._root_hdf5_location
+
+    @hdf_save_location.setter
+    def hdf_save_location(self, path: str):
+        if not os.path.isdir(path):
+            raise AttributeError(
+                f"Could not set {self.name} HDF5 save location. Please provide an existing directory."
+            )
+        self._root_hdf5_location = path
+
+    @property
+    def n_columns(self):
+        return self.controls_information.PVs.n_col.get()
+
+    @property
+    def n_rows(self):
+        return self.controls_information.PVs.n_row.get()
+
+    @property
+    def n_bits(self):
+        return self.controls_information.PVs.n_bits.get()
+
+    @property
+    def resolution(self):
+        return self.controls_information.PVs.resolution.get()
+
+    @property
+    def last_save_filepath(self):
+        return self._last_save_filepath
+
+    def _generate_new_filename(self, extension: Optional[str] = ".h5") -> str:
```

Comment:
This feels a bit screen-specific to put into a utils space for now, but I can add an issue to rewrite this function so that it doesn't depend on self.name for example